Frameworks/QuartzCore/CALayer.mm

```diff
@@ -518,69 +525,73 @@ - (void)display {
-        } else {
-            CGContextClearToColor(drawContext, 0, 0, 0, 0);
+                CGRect wholeRect = CGRectMake(0, 0, width, height);
+                CGContextFillRect(drawContext, wholeRect);
+                CGContextRestoreGState(drawContext);
+            }
 
-            CGContextSaveGState(drawContext);
-            CGContextSetFillColorWithColor(drawContext, [static_cast<UIColor*>(priv->_backgroundColor) CGColor]);
+            // UIKit and CALayer consumers expect the origin to be in the top left.
+            // CoreGraphics defaults to the bottom left, so we must flip and translate the canvas.
+            CGContextTranslateCTM(drawContext, 0, heightInPoints);
+            CGContextScaleCTM(drawContext, 1.0f, -1.0f);
+            CGContextTranslateCTM(drawContext, -priv->bounds.origin.x, -priv->bounds.origin.y);
 
-            CGRect wholeRect = CGRectMake(0, 0, width, height);
-            CGContextFillRect(drawContext, wholeRect);
-            CGContextRestoreGState(drawContext);
-        }
+            _CGContextSetShadowProjectionTransform(drawContext, CGAffineTransformMakeScale(1.0, -1.0));
 
-        // UIKit and CALayer consumers expect the origin to be in the top left.
-        // CoreGraphics defaults to the bottom left, so we must flip and translate the canvas.
-        CGContextTranslateCTM(drawContext, 0, heightInPoints);
-        CGContextScaleCTM(drawContext, 1.0f, -1.0f);
-        CGContextTranslateCTM(drawContext, -priv->bounds.origin.x, -priv->bounds.origin.y);
+            [self drawInContext:drawContext];
 
-        _CGContextSetShadowProjectionTransform(drawContext, CGAffineTransformMakeScale(1.0, -1.0));
+            if (priv->delegate != 0) {
+                if ([priv->delegate respondsToSelector:@selector(displayLayer:)]) {
+                    [priv->delegate displayLayer:self];
+                } else {
+                    [priv->delegate drawLayer:self inContext:drawContext];
+                }
+            }
 
-        CGContextSetDirty(drawContext, false);
-        [self drawInContext:drawContext];
+            _CGContextPopEndDraw(drawContext);
 
-        if (priv->delegate != 0) {
-            if ([priv->delegate respondsToSelector:@selector(displayLayer:)]) {
-                [priv->delegate displayLayer:self];
-            } else {
-                [priv->delegate drawLayer:self inContext:drawContext];
+            woc::StrongCF<CFErrorRef> renderError;
+            if (CGContextIwGetError(drawContext, &renderError)) {
+                switch (CFErrorGetCode(renderError)) {
+                    case kCGContextErrorDeviceReset:
+                        NSTraceInfo(TAG, @"Hardware device disappeared when rendering %@; retrying.", self);
+                        ++tries;
+                        continue;
+                    default: {
+                        NSTraceError(TAG, @"Error %@ rendering %@.", renderError.get(), self);
+                        break;
```

Comment:
Fail_fast here instead.  It would give us enough to up the number of retries. Another thing to note is that we should probably have another counter that falls back to warp if the particular device fails many times.  That is, if each frame takes 2 attempts to draw, we are probably much better off using warp.  This of course isn't necessary until we move to h/w acceleration.